### PR TITLE
UHM-4451: Config: Dynamically load all JavaScript and CSS in "scripts…

### PR DESCRIPTION
**Pull request by mogoodrich**
…" and "styles" directory

### CODE REVIEW

File: omod/src/main/webapp/pages/visit/visit.gsp [hunk 1]
```diff
@@ -37,8 +37,8 @@
     ui.includeJavascript("pihcore", "visit/vaccinations.js")
     ui.includeJavascript("pihcore", "visit/visit.js")
     ui.includeJavascript("pihcore", "visit/encounterTransaction.js")
-    ui.includeJavascript("pihcore", "visit/encounterTypeConfig.js")
     ui.includeJavascript("pihcore", "services/configService.js")
+    ui.includeJavascript("file", "configuration/pih/scripts/visit/encounterTypeConfig.js", null, true)
```
Comment (mogoodrich) [inline omod/src/main/webapp/pages/visit/visit.gsp:41]: See https://issues.openmrs.org/browse/UIFR-214 for the explanation behind "null" and "true"

File: api/src/main/java/org/openmrs/module/pihcore/setup/GlobalResourceSetup.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+            List<String> scriptResources = getGlobalResourcesFromFileSystemPath(PihCoreUtil.getGlobalScriptsDirectory());
+
+            for (String resource : scriptResources) {
+                addGlobalResource(Resource.CATEGORY_JS, "file", resource);
+            }
+        }
+        // this entire catch is a hack to get component test to pass until we find the proper way to mock this (see HtmlFormSetup where we do something similar)
+        catch (Exception e) {
+            // this is a hack to get component test to pass until we find the proper way to mock this
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/pihcore/setup/GlobalResourceSetup.java:43]:
```suggestion
        // this entire catch is a hack to get component test to pass until we find the proper way to mock this
        // (see HtmlFormSetup where we do something similar)
        catch (Exception e) {
```

File: api/src/main/java/org/openmrs/module/pihcore/setup/GlobalResourceSetup.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    protected static List<String> getGlobalResourcesFromFileSystemPath(String path) {
+
+        Collection<File> files = null;
+        List<String> filePaths = new ArrayList<>();
+
+        try {
+            File dir = new File(path);
+            if (dir.exists()) {
+                files = FileUtils.listFiles(dir, null, true);
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/pihcore/setup/GlobalResourceSetup.java:60]:
```suggestion
                files = FileUtils.listFiles(dir, /*extensions=*/ null,  /*recursive=*/ true);
```

File: api/src/main/java/org/openmrs/module/pihcore/setup/GlobalResourceSetup.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                filePaths.add(stripApplicationDataDirectory(file.getAbsolutePath()));
+            }
+        }
+
+        return filePaths;
+    }
+
+    // strip the application data directory from the path, returning the path relative to that directory
+    protected static String stripApplicationDataDirectory(String path) {
+        return path.replace(OpenmrsUtil.getApplicationDataDirectory(),"");
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/pihcore/setup/GlobalResourceSetup.java:78]:
```suggestion
        return path.replace(OpenmrsUtil.getApplicationDataDirectory(), "");
```

File: api/src/main/java/org/openmrs/module/pihcore/setup/GlobalResourceSetup.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    protected static void addGlobalResource(String category, String providerName, String resourcePath)  {
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/pihcore/setup/GlobalResourceSetup.java:81]:
```suggestion
    protected static void addGlobalResource(String category, String providerName, String resourcePath) {
```

File: api/src/main/java/org/openmrs/module/pihcore/setup/GlobalResourceSetup.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        ResourceFactory resourceFactory = ResourceFactory.getInstance();
+        PageFactory pageFactory = Context.getRegisteredComponents(PageFactory.class).get(0);
+        File resource = resourceFactory.getResource(providerName, resourcePath);
+        if (resource != null) {
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/pihcore/setup/GlobalResourceSetup.java:85]: Should you also check that `pageFactory != null`?

File: api/src/main/java/org/openmrs/module/pihcore/setup/GlobalResourceSetup.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+            GlobalResourceIncluder globalResourceIncluder = new GlobalResourceIncluder();
+            globalResourceIncluder.addResource(new Resource(category, providerName, resourcePath, -100));
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/pihcore/setup/GlobalResourceSetup.java:87]:
```suggestion
            int priority = -100;
            globalResourceIncluder.addResource(new Resource(category, providerName, resourcePath, priority));
```

Just for documentation. `/*priority=*/ -100` would also work fine.

File: omod/src/main/webapp/pages/visit/visit.gsp (hunk 1 above)
Comment (brandones) [reply omod/src/main/webapp/pages/visit/visit.gsp:41]: In keeping with my suggestions above, I'd inline that information into comments.

**Pull request merged**